### PR TITLE
[application] Add resnet18 tensorflow + update application @open sesame 09/23 13:59

### LINKED ISSUE #904: Evaluate ResNet

### LINKED ISSUE #1533: Enable resnet application
Resnet application fails to run:
- solve the fail reason
- ensure that resnet is run with each (with batch size 1 and single iteration)

This is need to verify #1127.

### PR DESCRIPTION
**Pull request by kparichay**
- Add resnet18 tensorflow version to match against nntrainer.
- Update resnet application to match tensorflow implementation.
- This patch reduces the number of operations required for tensor sum by
accumulating the consecutive axes in order to increase the precision of
the operation.
- This patch adds test check on the training of the resnet application.
This test can be enabled with enable-long-test and enable-test.

Note: reduction of operations for sum was intended to reduce the number of operations to result in higher precision but can very well result in speedup as well.

resolves #904
Resolves #1533

Signed-off-by: Parichay Kapoor <pk.kapoor@samsung.com>

### CODE REVIEW

File: Applications/Resnet/tensorflow/transLayer.py [hunk 1]
```diff
@@ -0,0 +1,200 @@
+#!/usr/bin/env python3
```
Comment (zhoonit) [inline Applications/Resnet/tensorflow/transLayer.py:1]: Could you use `sys.path.insert(1, {gen_input directory} )`; import transLayer as TL rather than including this file?

File: nntrainer/tensor/tensor.cpp [hunk 2]
```diff
@@ -738,6 +740,14 @@ Tensor Tensor::sum(const std::vector<unsigned int> &axes, float alpha) const {
   return sum(axes, ret, alpha);
 }
 
+void Tensor::merge_axis(unsigned int axis1, unsigned int axis2) {
```
Comment (zhoonit) [inline nntrainer/tensor/tensor.cpp:743]: maybe `mergeAxis`?

File: nntrainer/tensor/tensor.cpp [hunk 3]
```diff
@@ -738,6 +740,14 @@ Tensor Tensor::sum(const std::vector<unsigned int> &axes, float alpha) const {
...
+  if (axis2 != axis1 + 1)
+    throw std::invalid_argument("Axis to be merged must be continuous.");
```
Comment (zhoonit) [inline nntrainer/tensor/tensor.cpp:745]: is it also considered continuous if axis1 > axis2 ?

File: nntrainer/tensor/tensor.cpp [hunk 4]
```diff
@@ -738,6 +740,14 @@ Tensor Tensor::sum(const std::vector<unsigned int> &axes, float alpha) const {
...
+
+  dim.setTensorDim(axis2, dim.getTensorDim(axis1) * dim.getTensorDim(axis2));
+  dim.setTensorDim(axis1, 1);
```
Comment (zhoonit) [inline nntrainer/tensor/tensor.cpp:748]: Can we call Tensor::reshape instead?

File: nntrainer/tensor/tensor.cpp [hunk 5]
```diff
@@ -746,12 +756,22 @@ Tensor &Tensor::sum(const std::vector<unsigned int> &axes, Tensor &output,
   if (axes.size() == 1) {
     this->sum(axes[0], output, alpha);
   } else {
-    Tensor ret = this->sum(axes[0], alpha);
+    /** club axes together */
+    Tensor new_reshaped = *this;
+    std::vector<unsigned int> new_axes = {axes[0]};
+    for (unsigned int i = 1; i < axes.size(); ++i) {
+      if (axes[i] == axes[i - 1] + 1) {
```
Comment (zhoonit) [inline nntrainer/tensor/tensor.cpp:763]: note) it is not guaranteed that axes are always in sorted order. This is just an optimization so I think it's fine though

File: nntrainer/tensor/tensor.cpp [hunk 6]
```diff
@@ -746,12 +756,22 @@ Tensor &Tensor::sum(const std::vector<unsigned int> &axes, Tensor &output,
...
+        new_reshaped.merge_axis(axes[i - 1], axes[i]);
+        new_axes.back() = axes[i];
+      } else {
+        new_axes.push_back(axes[i]);
+      }
+    }
 
-    for (unsigned int i = 1; i < axes.size() - 1; ++i)
+    Tensor ret = new_reshaped.sum(new_axes[0]);
```
Comment (zhoonit) [inline nntrainer/tensor/tensor.cpp:771]: wonder if this does not conflict with line 757?

File: Applications/Resnet/tensorflow/transLayer.py (hunk 1 above)
Comment (kparichay) [reply Applications/Resnet/tensorflow/transLayer.py:1]: done.

File: nntrainer/tensor/tensor.cpp (hunk 6 above)
Comment (zhoonit) [inline nntrainer/tensor/tensor.cpp:771]: we might need more unittest for Tensor::sum

File: nntrainer/tensor/tensor.cpp (hunk 2 above)
Comment (kparichay) [reply nntrainer/tensor/tensor.cpp:743]: done.

File: nntrainer/tensor/tensor.cpp (hunk 3 above)
Comment (kparichay) [reply nntrainer/tensor/tensor.cpp:745]: Update the error message to be more precise. We only merge is axis2 == axis1 + 1.

File: Applications/Resnet/tensorflow/transLayer.py (hunk 1 above)
Comment (zhoonit) [reply Applications/Resnet/tensorflow/transLayer.py:1]: I guess this file can be deleted!

File: nntrainer/tensor/tensor.cpp (hunk 4 above)
Comment (kparichay) [reply nntrainer/tensor/tensor.cpp:748]: To use reshape, there has to be a switch making this slower in critical section depending on the location of axis.

File: nntrainer/tensor/tensor.cpp (hunk 3 above)
Comment (zhoonit) [reply nntrainer/tensor/tensor.cpp:745]: Cool

**Review by zhoonit — APPROVED** [review]
LGTM :)

File: nntrainer/tensor/tensor.cpp (hunk 5 above)
Comment (kparichay) [reply nntrainer/tensor/tensor.cpp:763]: Yes, its not gauranteed. Developer must try to give it in order for optimization. We can add sort and unique but IMO its overkill.

File: nntrainer/tensor/tensor.cpp (hunk 6 above)
Comment (kparichay) [reply nntrainer/tensor/tensor.cpp:771]: I will add an issue for this.

File: nntrainer/tensor/tensor.cpp (hunk 4 above)
Comment (zhoonit) [reply nntrainer/tensor/tensor.cpp:748]: If we have change the reshape implementation, this will be likely to be untouched. If we have more detailed tests, I guess it will be fine :)

File: Applications/Resnet/tensorflow/transLayer.py (hunk 1 above)
Comment (kparichay) [reply Applications/Resnet/tensorflow/transLayer.py]: it is :).

**Review by jijoongmoon — APPROVED** [review]
LGTM

File: nntrainer/tensor/tensor.cpp (hunk 6 above)
Comment (kparichay) [reply nntrainer/tensor/tensor.cpp:771]: #1581

File: nntrainer/tensor/tensor.cpp (hunk 4 above)
Comment (kparichay) [reply nntrainer/tensor/tensor.cpp:748]: Yes, will include tests for this with #1581.
